devito/types.py

```diff
@@ -397,56 +418,86 @@ def size(self):
-        return AbstractCachedFunction.__new__(cls, *args, **kwargs)
+        return EnrichedTuple(*offsets, getters=self.dimensions, left=left, right=right)
 
-    def update(self):
-        return
+    @property
+    def _offset_halo(self):
+        """
+        The number of grid points between the first (last) allocated element
+        (possibly in the halo/padding region) and the first (last) halo element,
+        for each dimension.
+        """
+        left = self._extent_padding.left
+        right = self._extent_padding.right
 
+        Offset = namedtuple('Offset', 'left right')
+        offsets = tuple(Offset(i, j) for i, j in self._padding)
 
-class Scalar(Symbol):
-    """Symbolic object representing a scalar.
+        return EnrichedTuple(*offsets, getters=self.dimensions, left=left, right=right)
 
-    :param name: Name of the symbol
-    :param dtype: Data type of the scalar
-    """
+    @property
+    def _extent_halo(self):
+        """
+        The number of grid points in the halo region.
+        """
+        left = tuple(zip(*self._halo))[0]
+        right = tuple(zip(*self._halo))[1]
 
-    is_Scalar = True
+        Extent = namedtuple('Extent', 'left right')
+        extents = tuple(Extent(i, j) for i, j in self._halo)
 
-    def __init__(self, *args, **kwargs):
-        if not self._cached():
-            self.dtype = kwargs.get('dtype', np.float32)
+        return EnrichedTuple(*extents, getters=self.dimensions, left=left, right=right)
 
     @property
-    def _mem_stack(self):
-        """Return True if the associated data should be allocated on the stack
-        in a C module, False otherwise."""
-        return True
+    def _extent_padding(self):
+        """
+        The number of grid points in the padding region.
+        """
+        left = tuple(zip(*self._padding))[0]
+        right = tuple(zip(*self._padding))[1]
 
-    def update(self, dtype=None, **kwargs):
-        self.dtype = dtype or self.dtype
+        Extent = namedtuple('Extent', 'left right')
+        extents = tuple(Extent(i, j) for i, j in self._padding)
 
+        return EnrichedTuple(*extents, getters=self.dimensions, left=left, right=right)
 
-class Array(SymbolicData):
-    """Symbolic object representing a tensor.
+    @property
+    def _mask_domain(self):
+        """A mask to access the domain region of the allocated data."""
+        return tuple(slice(i, -j) if j != 0 else slice(i, None)
+                     for i, j in self._offset_domain)
 
-    :param name: Name of the symbol
-    :param dtype: Data type of the scalar
-    :param shape: The shape of the tensor
-    :param dimensions: The symbolic dimensions of the tensor.
+    @property
+    def _mask_with_halo(self):
+        """A mask to access the domain+halo region of the allocated data."""
+        return tuple(slice(i, -j) if j != 0 else slice(i, None)
+                     for i, j in self._offset_halo)
+
+
+class Array(AbstractCachedFunction):
```

Comment:
Not a big fan of the name, would prefer something like `FunctionValue` o

Reply:
u mean ``Array``? it's been in ``master`` for ages. Just changed its place